### ISSUE TITLE
Allow to use the built-in encryption method as a secret store

### ISSUE THREAD

**Issue by infused-kim** [twpayne/chezmoi#3537]
First of all I want to compliment you on thank you for creating an absolutely incredible piece of software. You really did think of everything and created something incredibly powerful.

I highly appreciate you commitment and work not only to create it, but also to maintain it and provide support for it.

## Is your feature request related to a problem? Please describe.

I want to deploy my chezmoi managed dotfiles on servers that I don't control and don't fully trust. Unfortunately, I also need to use some secrets on these servers that I wouldn't want to store in my git repo and of course I don't want to expose my actual password manager database to an untrusted server.

So, ideally I would like to store the required secrets encrypted in the dotfiles repo and not need an additional dependency or software to access them.

## Describe the solution you'd like

I would like an additional secrets / password manager backend that uses the built-in encryption to access the secrets.

This is already possible (described below as a workaround in the meantime), but it's a little clunky and it requests the password every time a secret is accessed (secrets are not cached unlike with the generic secrets command).

## Describe alternatives you've considered

Here is how I ended up implementing it. Make sure to do this on a trusted host.

### 1. Enable encryption and define the secrets file in chezmoi
Edit `.local/share/chezmoi/.chezmoi.toml.tmpl`:
```toml
{{ $passphrase := promptStringOnce . "passphrase" "encryption passphrase (it will be stored in the config if not empty)" -}}

{{ $should_ignore_encrypted := promptBoolOnce . "should_ignore_encrypted" "should exclude encrypted files" -}}

[data]
    secrets_path = {{ joinPath .chezmoi.sourceDir "dot_config/chezmoi/encrypted_secrets.toml.asc" | quote }}

    should_ignore_encrypted = {{ $should_ignore_encrypted }}

[diff]
    pager = "delta"

encryption = "gpg"

[gpg]
    symmetric = true
{{ if ne $passphrase "" }}
    args = ["--batch", "--passphrase", {{ $passphrase | quote }}, "--no-symkey-cache"]
{{ end }}
```

### 2. Create a file that holds all the secrets
Create `.config/chezmoi/secrets.toml`:
```toml
secret = "secret secrets are no fun"

another = "secret secrets hurt someone"
```

Then encrypt and add it to chezmoi with: `chezmoi add --encrypt .config/chezmoi/secrets.toml`

### Use the secrets in config files:
For example in `.local/share/chezmoi/test_secrets.txt`:
```txt
{{- $secret := (fromToml (.secrets_path | include | decrypt)).secret -}}
{{- $another := (fromToml (.secrets_path | include | decrypt)). another -}}

My secret: {{ $secret }}
And another: {{ $another }}
```

### Exclude encrypted files

Chezmoi asks to enter the encryption password for each file and each time a secret is accessed. It also doesn't show which file it is decrypting.

I can think of many attack vectors, on servers that are not controlled by you, where secrets could get leaked that shouldn't have been decrypted.

Most of my configs that require secrets don't change often, so I prefer a workflow where only on initial setup the files are generated. But on subsequent updates, I don't want to be asked for the password and risk decrypting something I didn't mean to decrypt.

I achieve this with `should_ignore_encrypted` in the config. When it is set to true, I ignore all encrypted files in `.chezmoiignore`:

```gitignore
# Don't store the decrypted secrets ever
.config/chezmoi/secrets.toml

{{ if eq .should_ignore_encrypted true -}}
.ssh/id_rsa
.config/config_with_secrets/conf
{{ end }}
```

### Add or remove secrets

On a trusted machine comment out `.config/chezmoi/secrets.toml` in the `.chezmoiignore` and run `chezmoi apply` to produce an unencrypted version of the file.

Edit it and then add it back to chezmoi with `chezmoi forget .config/chezmoi/secrets.toml && chezmoi add --encrypt .config/chezmoi/secrets.toml`.

**Comment by twpayne** [comment]
tl;dr use [age](https://github.com/FiloSottile/age)'s support for multiple recipients instead of GPG.

Thank you for your kind words and for your extremely well-thought-through issue! Really appreciated :)

I would suggest an alternative solution that should meet your requirements.

More info:

Firstly, secrets management is hard. This is why chezmoi delegates all secret management to password managers and/or established encryption software. I am personally definitely unable to design the protocols or implement the code without fault.

For this specific case, I would use the following approach:

1. Generate an `age` private key (recipient) for each machine (or group of machines) which you only want to have access to limited secrets, even if that machine is completely compromised.
2. Encrypt each secret so that it can only be read by the [specific recipient](https://github.com/FiloSottile/age?tab=readme-ov-file#multiple-recipients) that you want it readable on.
3. Use [this part of the user guide](https://www.chezmoi.io/user-guide/frequently-asked-questions/encryption/#how-do-i-configure-chezmoi-to-encrypt-files-but-only-request-a-passphrase-the-first-time-chezmoi-init-is-run) to set up the recipient on each machine.

This way, you only have to enter a password once per machine, and even if an individual machine is completely compromised, the rest of your secrets should be safe.

**Comment by infused-kim** [comment]
Thank you! That's a really great solution.

**Comment by twpayne** [comment]
Hopefully this is now resolved. Please re-open if needed.
